File: Algorithm_179_Largest Number.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
using namespace std;
class Solution
{
  public:
    string largestNumber(vector<int> &nums)
    {
        if (nums.size() == 0) return "";
        vector<string> numsStr;
        string res;
        for (int i = 0; i < nums.size() ; i++) numsStr.push_back(to_string(nums[i]));
        sort(numsStr.begin(), numsStr.end(), myCompare);
        for (int i = 0; i < numsStr.size() ; i++) res += numsStr[i];
        int start;
        for (start = 0; res[start] == '0' ; start++);
        res = res.substr(start, res.size() - start);
        return res.size() > 0 ? res : "0";
    }
    struct compare {
        bool operator() (string &s1, string &s2) {return s1 + s2 > s2 + s1;}
    } myCompare;
};

File: Algorithm_179_Largest Number.java
import java.util.Arrays;

// 思路：把nums中的int全部转换为String，然后排序，排序规则为对于两字符串s1、s2，若s1 + s2 > s2 + s1，则s1 + s2排在前面
// 例如s1 = "85", s2 = "9", s1 + s2 = "829", s2 + s1 = "985", 所以s2 + s1排在s1 + s2前面
// 排序后，把所有String按顺序连接起来，就是所要求的结果

class Solution {
    public String largestNumber(int[] nums) {
        if (nums == null || nums.length == 0) {
            return "0";
        }
        StringBuilder res = new StringBuilder();
        String[] numsStr = new String[nums.length];
        for (int i = 0; i < nums.length; i++) {
            numsStr[i] = String.valueOf(nums[i]);
        }
        Arrays.sort(numsStr, (s1, s2) -> (s2 + s1).compareTo(s1 + s2)); // sort by descending order
        if (numsStr[0].charAt(0) == '0') { // the largest num in nums is 0, then return "0"
            return "0";
        }
        for (String num : numsStr) {
            res.append(num);
        }
        return res.toString();
    }
}